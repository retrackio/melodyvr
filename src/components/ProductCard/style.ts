import styled from 'styled-components';

interface WrapperProps {
  accentColor: string;
  image: string;
}

interface TitleProps {
  backgroundColor: string;
}

export const Container = styled.div`
  position: relative;
`;

export const Link = styled.a`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  z-index: 2;
`;

export const Wrapper = styled.div<WrapperProps>`
  box-sizing: initial;
  height: 250px;
  width: 100%;
  overflow: hidden;
  position: relative;
  border-top: 6px solid ${(props) => `#${props.accentColor}`};
  background-image: ${(props) => `url(${props.image})`};
  background-size: cover;
  background-position: 50% 50%;
`;

export const Title = styled.h2<TitleProps>`
  font-family: 'Proza Libre';
  font-weight: 600;
  font-size: 1.8rem;
  color: #fff;
  position: absolute;
  left: 0px;
  top: 0px;
  padding: 10px;
  background-color: ${(props) => `#${props.backgroundColor}`};
  box-shadow: 2px 2px 12px rgba(0, 0, 0, 0.4);
`;

export const Description = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  bottom: 0;
  padding: 10px;
  line-height: 1.8rem;
  color: #fff;
  background: rgba(0, 0, 0, 0.5);
  font-size: 1.3rem;
`;
