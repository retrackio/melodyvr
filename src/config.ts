if (typeof window === 'undefined') {
  require('dotenv').config();
}

export default {
  app: {
    env: process.env.NODE_ENV,
    port: parseInt(process.env.PORT, 10) || 3000,
    logLevel: process.env.LOG_LEVEL || 'info',
    slugFormat: {
      lower: true,
      replacement: '_',
      strict: true,
    },
  },
  services: {
    apiBaseUrl: process.env.API_BASE_URL,
  },
};
