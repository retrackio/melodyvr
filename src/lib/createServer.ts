import express from 'express';
import Server from 'next/dist/next-server/server/next-server';

export default function createServer(app: Server) {
  const server = express();
  const handle = app.getRequestHandler();

  server.all(
    '*',
    (req: express.Request, res: express.Response) => handle(req, res),
  );

  return server;
}
