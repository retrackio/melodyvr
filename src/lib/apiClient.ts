import fetch from 'isomorphic-fetch';
import config from '../config';

export interface ReleasesResult {
  result: {
    releases: Release[];
  };
}

export interface Release {
  feature_order: number;
  is_free: boolean;
  product: Product;
}

export interface Product {
  id: number;
  title: string;
  description: string;
  image_hero_url: string;
  image_landscape_url: string;
  accent_colour_code: string;
  background_colour_code: string;
  text_colour_code: string;
  production_type_code: string;
  price: string | undefined;
}

export interface Venue {
  id: number;
  name: string;
  description: string;
  city: string;
  country_code: string;
}

export interface Artist {
  id: number;
  name: string;
  description: string;
  country_code: string;
  website_url: string;
  facebook_url: string;
  instagram_url: string;
  twitter_url: string;
  image_url: string;
  artist_background_colour_code: string;
  artist_text_colour_code: string;
  artist_accent_colour_code: string;
}

export interface Genre {
  id: number;
  name: string;
  description: string | null;
}

export type Performance = Product & {
  venue: Venue;
  artist: Artist;
  genre: Genre;
};

function sortReleases(releases: Release[]) {
  return releases.sort(
    (releaseA, releaseB) => releaseA.feature_order - releaseB.feature_order,
  );
}

export async function getReleases(): Promise<Release[]> {
  const response = await fetch(`${config.services.apiBaseUrl}/products`);
  const {
    result: { releases },
  } = await response.json();
  return sortReleases(releases);
}

export async function getPerformance(id: number): Promise<Performance> {
  const response = await fetch(`${config.services.apiBaseUrl}/products/${id}`);
  const { result } = await response.json();
  return result;
}
