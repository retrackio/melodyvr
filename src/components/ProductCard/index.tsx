import React from 'react';
import Link from 'next/link';
import slugify from 'slugify';
import { Product } from '../../lib/apiClient';
import config from '../../config';
import ProductPrice from '../ProductPrice';
import * as Styled from './style';

interface Props {
  product: Product;
  isFree: boolean;
}

const ProductCard = ({ product, isFree }: Props) => (
  <Styled.Container>
    <Link
      href="/performance/[id]/[slug]"
      as={`/performance/${product.id}/${slugify(
        product.title,
        config.app.slugFormat,
      )}`}
      passHref
    >
      <Styled.Link />
    </Link>
    <Styled.Wrapper
      accentColor={product.accent_colour_code}
      image={product.image_hero_url}
    >
      <Styled.Title backgroundColor={product.accent_colour_code}>
        {product.title}
      </Styled.Title>
      <ProductPrice price={product.price} isFree={isFree} />
      <Styled.Description
        dangerouslySetInnerHTML={{ __html: product.description }}
      />
    </Styled.Wrapper>
  </Styled.Container>
);

export default ProductCard;
