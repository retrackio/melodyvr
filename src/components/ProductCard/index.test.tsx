import React from 'react';
import { shallow } from 'enzyme';
import slugify from 'slugify';
import ProductCard from '.';
import { Product } from '../../lib/apiClient';
import config from '../../config';

jest.mock('slugify', () => jest.fn().mockReturnValue('some_slug'));

const mockedProduct: Product = {
  id: 2,
  title: 'Imagine Dragons',
  description: '<i>Performance contains flashing or strobe lighting</i>.',
  image_hero_url: 'hero_image.jpg',
  image_landscape_url: 'landscape_image.jpg',
  accent_colour_code: 'F52459',
  background_colour_code: '1e1e1e',
  text_colour_code: 'ffffff',
  production_type_code: 'performance',
  price: '9.99',
};

describe('<ProductCard />', () => {
  afterAll(() => jest.resetAllMocks());

  it('renders correctly', () => {
    const wrapper = shallow(
      <ProductCard product={mockedProduct} isFree={false} />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('generates the correct slug for the url', () => {
    shallow(<ProductCard product={mockedProduct} isFree={false} />);
    expect(slugify).toHaveBeenCalledWith(
      mockedProduct.title,
      config.app.slugFormat,
    );
  });

  it('links to the correct page', () => {
    const wrapper = shallow(
      <ProductCard product={mockedProduct} isFree={false} />,
    );
    const link = wrapper.find('Link');
    expect(link.prop('href')).toEqual('/performance/[id]/[slug]');
    expect(link.prop('as')).toEqual('/performance/2/some_slug');
  });
});
