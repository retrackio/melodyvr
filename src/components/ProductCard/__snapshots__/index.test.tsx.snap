// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> renders correctly 1`] = `
<style__Container>
  <Link
    as="/performance/2/some_slug"
    href="/performance/[id]/[slug]"
    passHref={true}
  >
    <style__Link />
  </Link>
  <style__Wrapper
    accentColor="F52459"
    image="hero_image.jpg"
  >
    <style__Title
      backgroundColor="F52459"
    >
      Imagine Dragons
    </style__Title>
    <ProductPrice
      isFree={false}
      price="9.99"
    />
    <style__Description
      dangerouslySetInnerHTML={
        Object {
          "__html": "<i>Performance contains flashing or strobe lighting</i>.",
        }
      }
    />
  </style__Wrapper>
</style__Container>
`;
